### DISCUSSION TITLE
How to safely iterate though a list that may be updated on another thread?

### DISCUSSION THREAD

**Discussion by benstevens48** [microsoft/WindowsAppSDK#4442] [Q&A]
Many of the WinRT APIs return an IVector or even IVectorView, which are projected as IList and IReadOnlyList in C#. Some of these are subject to being updated on background threads (even the read-only list), for example MediaPlaybackItem.TimedMetadataTracks. How is one meant to safely iterate through such a collection? For example, I just tried iterating though MediaPlaybackItem.TimedMetadataTracks using a foreach loop in C# and got a collection modified exception (and I am not modifying it in the loop). I guess I can do ToList first or something, but presumably that would just reduce the chance of such an exception by reducing the time spent iterating, not remove it completely (unless it uses some other internal method to get the elements from the Vector).

Is the only solution to wrap the loop in a try-catch block? In this case it would be good if the docs called out all instances where this is necessary.

**Comment by kmgallahan** [comment]
https://learn.microsoft.com/en-us/uwp/api/windows.media.playback.mediaplaybacktimedmetadatatracklist?view=winrt-22621#remarks

> Get an instance of this class by accessing the [TimedMetadataTracks](https://learn.microsoft.com/en-us/uwp/api/windows.media.playback.mediaplaybackitem.timedmetadatatracks?view=winrt-22621#windows-media-playback-mediaplaybackitem-timedmetadatatracks) property of a [MediaPlaybackItem](https://learn.microsoft.com/en-us/uwp/api/windows.media.playback.mediaplaybackitem?view=winrt-22621). To be notified when the list of tracks changes, handle the [TimedMetadataTracksChanged](https://learn.microsoft.com/en-us/uwp/api/windows.media.playback.mediaplaybackitem.timedmetadatatrackschanged?view=winrt-22621#windows-media-playback-mediaplaybackitem-timedmetadatatrackschanged) event.

Maybe exit the loop via a cancelation token-like mechanism if the event is raised during enumeration. I don't know how reliable the timing of the event is though.

**Reply by benstevens48** [reply]
That doesn't really help since the code is running on different threads there's no guarantee what order it will run in. Seems like maybe not great design? Or maybe try-catch is standard in this sort of scenario? Not sure if many if any devs would realise that though.

**Reply by kmgallahan** [reply]
>  I guess I can do ToList first or something, but presumably that would just reduce the chance of such an exception by reducing the time spent iterating, not remove it completely

Maybe try using `MediaPlaybackTimedMetadataTrackList.GetMany(UInt32, TimedMetadataTrack[])` to just get all of them at once if it isn't expensive.

**Reply by benstevens48** [reply]
Is that available in C#?

**Reply by kmgallahan** [reply]
I'm just looking at the docs for the classes you mentioned:

https://learn.microsoft.com/en-us/uwp/api/windows.media.playback.mediaplaybacktimedmetadatatracklist?view=winrt-22621

**Reply by benstevens48** [reply]
I can't test it now but I don't think that method is projected in C#.
